### ISSUE TITLE
Skeletons shaking

### ISSUE THREAD

**Issue by Zinibow** [ruvnet/wifi-densepose#288]
Hello, I am new to this, and I am still trying to figure out the best settings. I do not have a model, and I have tried training, but when I run curl -X POST http://localhost:3000/api/v1/adaptive/train or curl http://localhost:3000/api/v1/adaptive/status it instantly goes to the next line, and I can type again, meaning it instantly quits. I spent an hour debugging it and trying to run it with a PowerShell command, but nothing. I have 3 devices around my room (but multiple people and animals in the house), and it is extremely shaky and barely knows where people are. I have a high enough accuracy according to your guide. The confidence level is about 50. I run the server with docker run -p 3000:3000 -p 3001:3001 -p 5005:5005/udp -e CSI_SOURCE=esp32 ruvnet/wifi-densepose:latest. I also don't see an observatory category. Is there a video tutorial out there anywhere? Thanks!

My discord is talon2 if you would rather try and help me there, so we don't have to go back and forth in an issue post, thanks!

**Comment by ruvnet** [comment]
## Why training returns instantly

The `POST /api/v1/adaptive/train` endpoint loads labeled `.csi.jsonl` files from `data/recordings/`. If that directory is empty or doesn't exist, it returns immediately with no training data.

**You need to record labeled data first:**

```bash
# Record ~60s of empty room
curl -X POST "http://localhost:3000/api/v1/record/start?label=empty"
# wait 60s
curl -X POST "http://localhost:3000/api/v1/record/stop"

# Record ~60s with you sitting still
curl -X POST "http://localhost:3000/api/v1/record/start?label=still"
# wait 60s
curl -X POST "http://localhost:3000/api/v1/record/stop"

# Record ~60s with you walking around
curl -X POST "http://localhost:3000/api/v1/record/start?label=walking"
# wait 60s
curl -X POST "http://localhost:3000/api/v1/record/stop"

# Now train — this should take a few seconds and return accuracy stats
curl -X POST http://localhost:3000/api/v1/adaptive/train
```

If using Docker, make sure you mount a volume for `data/recordings/` so the files persist.

## Why skeletons shake

The pose keypoints are **synthesized from CSI signal features** (variance, motion band power, dominant frequency), not from a neural network model. This means:

- Signal noise → skeleton jitter  
- Multiple people + animals in the house → noisy CSI → more shaking  
- With 3 nodes this should be improvable via adaptive training (see recording steps above)

After training, the adaptive model blends its confidence (70%) with the raw heuristic (30%), which significantly reduces noise for your specific room.

## Confidence of ~50

A confidence of 50% is on the boundary between 1 and 2 persons (threshold is at 0.50). This explains the instability — the system is oscillating around the decision boundary. Adaptive training will shift this for your environment.

**Comment by ruvnet** [comment]
## Update: recent fixes that should help

@Zinibow — a few things have shipped since this was reported:

### 1. Person count stability (PR #295)
The "extremely shaky" skeletons were partly caused by the person count estimator flipping rapidly between values. PR #295 added real hysteresis with a smoothing window — the count now requires consistent signal changes before updating. This should significantly reduce the jitter you're seeing.

### 2. Node offline detection (PR #300)
Nodes are now marked offline after 5 seconds of no frames. Previously, stale node data could cause phantom detections.

### 3. Observatory page
The observatory UI is at `http://localhost:3000/ui/observatory.html` when running the Rust `sensing-server`. If you're using the Docker image, make sure you're on a recent build — older Docker images may not include the observatory page.

```bash
# Pull latest Docker image
docker pull ruvnet/wifi-densepose:latest

# Run with all ports
docker run -p 3000:3000 -p 3001:3001 -p 5005:5005/udp \
  -e CSI_SOURCE=esp32 ruvnet/wifi-densepose:latest
```

### 4. Training requires recorded data first
As noted in the earlier reply, `POST /adaptive/train` returns instantly if there's no recorded data. The recording workflow is:

```bash
# Record empty room (~60s)
curl -X POST "http://localhost:3000/api/v1/record/start?label=empty"
# wait 60s
curl -X POST "http://localhost:3000/api/v1/record/stop"

# Record with 1 person (~60s)
curl -X POST "http://localhost:3000/api/v1/record/start?label=1person"
# wait 60s
curl -X POST "http://localhost:3000/api/v1/record/stop"

# Now train
curl -X POST "http://localhost:3000/api/v1/adaptive/train"
```

With 3 nodes and multiple people/animals in the house, you'll want to record calibration data with a controlled environment first (empty room, then just you).

**Comment by Zinibow** [comment]
I am unable to run any of the start recording commands; those also return instantly. I have tried running it in PowerShell as well. 

<img width="861" height="150" alt="Image" src="https://github.com/user-attachments/assets/41b9791b-3177-4da4-baeb-ad1b2bf0df50" />

<img width="1002" height="126" alt="Image" src="https://github.com/user-attachments/assets/2f24cecd-2e36-4d1b-80cc-429990f1c661" />

<img width="1019" height="522" alt="Image" src="https://github.com/user-attachments/assets/d5f2b620-8de8-45c8-ac13-c175a8f9c163" />

<img width="1730" height="826" alt="Image" src="https://github.com/user-attachments/assets/36609f9e-b690-46d2-9da3-155e6da5340f" />

**Comment by ruvnet** [comment]
## Fix: wrong API endpoint in previous instructions

@Zinibow — apologies, the curl commands in my earlier replies used the **wrong URL path**. The correct endpoints are:

```bash
# CORRECT — note "recording" not "record"
curl -X POST "http://localhost:3000/api/v1/recording/start?label=empty"
curl -X POST "http://localhost:3000/api/v1/recording/stop"

# WRONG (what was previously given)
# curl -X POST "http://localhost:3000/api/v1/record/start?label=empty"
```

The server returns an empty 404 for unknown routes, which is why curl exits instantly with no output.

### Full corrected calibration workflow

```bash
# 1. Record empty room (~60 seconds)
curl -X POST "http://localhost:3000/api/v1/recording/start?label=empty"
timeout 60 && curl -X POST "http://localhost:3000/api/v1/recording/stop"

# 2. Record you sitting still (~60 seconds)
curl -X POST "http://localhost:3000/api/v1/recording/start?label=still"
timeout 60 && curl -X POST "http://localhost:3000/api/v1/recording/stop"

# 3. Record you walking (~60 seconds)
curl -X POST "http://localhost:3000/api/v1/recording/start?label=walking"
timeout 60 && curl -X POST "http://localhost:3000/api/v1/recording/stop"

# 4. Verify recordings exist
curl http://localhost:3000/api/v1/recording/list

# 5. Train adaptive model
curl -X POST http://localhost:3000/api/v1/adaptive/train
```

### If using Docker

The Docker image does **not** mount a volume for `data/recordings/` by default, so recordings are lost when the container restarts. Add a volume mount:

```bash
docker run -p 3000:3000 -p 3001:3001 -p 5005:5005/udp \
  -v ruview-recordings:/app/data/recordings \
  -e CSI_SOURCE=esp32 ruvnet/wifi-densepose:latest
```

### Verify it's working

After `curl -X POST .../recording/start?label=empty`, you should see a JSON response like:
```json
{"status":"recording","id":"...","label":"empty"}
```

If you get an empty response or HTML, the endpoint isn't being reached — double-check the port and that the server is running.

Let me know if this unblocks you.

**Comment by ruvnet** [comment]
The shaky skeletons are expected behavior without a trained model — the system needs calibration for your specific environment.

Regarding the training API returning instantly: the `/api/v1/adaptive/train` endpoint starts training asynchronously. Check status with `curl http://localhost:3000/api/v1/adaptive/status` — it should show training progress. If it shows no model, you may need more CSI data collected first (let the system run for a few minutes with nodes active).

Tips:
- With 3 devices + multiple people/animals, there's a lot of signal interference — start with 1 person in the room during training
- See #301 for the setup troubleshooting guide

Closing as a support question. Feel free to reopen if the training endpoint genuinely fails with an error.
